tools/bpf2c/bpf_code_generator.cpp

```diff
@@ -354,62 +354,116 @@ bpf_code_generator::extract_program(const bpf_code_generator::unsafe_string& sec
-            std::string unsafe_symbol_name;
-            ELFIO::Elf64_Addr symbol_value{};
-            unsigned char symbol_bind{};
-            unsigned char symbol_type{};
-            ELFIO::Elf_Half symbol_section_index{};
-            unsigned char symbol_other{};
-            ELFIO::Elf_Xword symbol_size{};
-
-            symbols.get_symbol(
-                i,
-                unsafe_symbol_name,
-                symbol_value,
-                symbol_size,
-                symbol_bind,
-                symbol_type,
-                symbol_section_index,
-                symbol_other);
-
-            if (symbol_section_index == map_section->get_index()) {
-                if (symbol_size != sizeof(ebpf_map_definition_in_file_t)) {
-                    throw bpf_code_generator_exception("invalid map size");
-                }
-                if (symbol_value > map_section->get_size()) {
-                    throw bpf_code_generator_exception("invalid symbol value");
-                }
-                if ((symbol_value + symbol_size) > map_section->get_size()) {
-                    throw bpf_code_generator_exception("invalid symbol value");
-                }
+static std::tuple<std::string, ELFIO::Elf_Half>
+get_symbol_name_and_section_index(ELFIO::const_symbol_section_accessor& symbols, ELFIO::Elf_Xword index)
+{
+    std::string symbol_name;
+    ELFIO::Elf64_Addr value{};
+    ELFIO::Elf_Xword size{};
+    unsigned char bind{};
+    unsigned char type{};
+    ELFIO::Elf_Half section_index{};
+    unsigned char other{};
+    symbols.get_symbol(index, symbol_name, value, size, bind, type, section_index, other);
+    return {symbol_name, section_index};
+}
 
-                map_definitions[unsafe_symbol_name].definition =
-                    *reinterpret_cast<const ebpf_map_definition_in_file_t*>(map_section->get_data() + symbol_value);
+// We should consider refactoring the code that parses ELF files into a form that can be used by both ebpf-verifier and
+// bpf2c.
+void
+bpf_code_generator::parse_legacy_maps_section(const unsafe_string& name)
+{
+    auto map_section = get_optional_section(name);
+    if (!map_section) {
+        return;
+    }
 
-                map_definitions[unsafe_symbol_name].index = symbol_value / sizeof(ebpf_map_definition_in_file_t);
+    // Count the number of symbols that point into this maps section.
+    ELFIO::const_symbol_section_accessor symbols{reader, get_required_section(".symtab")};
+    int map_count = 0;
+    for (ELFIO::Elf_Xword index = 0; index < symbols.get_symbols_num(); index++) {
+        auto [symbol_name, section_index] = get_symbol_name_and_section_index(symbols, index);
+        if ((section_index == map_section->get_index()) && !symbol_name.empty())
+            map_count++;
+    }
+    if (map_count == 0) {
+        return;
+    }
+
+    size_t data_size = map_section->get_size();
+    size_t map_record_size = data_size / map_count;
+
+    if (data_size % map_record_size != 0) {
```

Comment:
Should we have a check before this line that `map_record_size != 0` ?

Reply:
Done